### DISCUSSION TITLE
OpenGL renderer support! Feedback wanted!

### DISCUSSION THREAD

**Discussion by rafaelvcaetano** [rafaelvcaetano/melonDS-android#1267] [General]
The OpenGL renderer has finally been introduced and, with it, comes the most requested feature since melonDS for Android began its development: **internal resolution scaling** or **high resolution rendering**. This allows the emulator to render the content using a resolution higher than the Nintendo DS's measly 256x192 resolution (per screen), which will give much more detail to 3D objects and scenes. To enable it, go to the Video settings and select the OpenGL renderer. If you can't see this option, it means that your device does not support OpenGL ES 3.2, which is required. You can scale the internal resolution up to 8 times, but I really doubt your device will handle it properly.

Here you can see a comparison between the software renderer and the OpenGL renderer running at 3x scaling:
| Software | OpenGL (3x) |
|---|---|
| ![image](https://github.com/rafaelvcaetano/melonDS-android/assets/9086379/8bf78173-955e-4045-bb21-9d5a4fcd1ecb) | ![image](https://github.com/rafaelvcaetano/melonDS-android/assets/9086379/5f000600-59c3-49ab-b49e-e721cdc5a703) |

## Performance
At 1x scaling, the performance is extremely similar to the software renderer. As you increase the resolution scaling, performance slightly goes down. To use use the highest scaling levels with acceptable performance, you will probably need a current gen flagship. But I don't have one, so I couldn't really test.

The current implementation is, above all, used as a proof that the renderer can run on Android. Some shortcuts were taken, which means that performance can still be slightly improved.

## Rendering issues
Just like in the PC version of melonDS, the OpenGL renderer is less accurate that the software one. As such, you are bound to find some visual artefacts or issues in some games. If you find such an issue, please take a look at [melonDS's OpenGL issues](https://github.com/melonDS-emu/melonDS/issues?q=is%3Aissue+is%3Aopen+label%3AOpenGL) to check if the PC version is affected by it as well. If it can't be reproduced on the PC version, feel free to open tickets so that I can take a look at them.

## How to test
Download the latest [nightly build](https://github.com/rafaelvcaetano/melonDS-android/releases/tag/nightly-release), enable the OpenGL render in the Video settings, and you are good to go! Don't forget to leave your feedback in this discussion!

**Comment by Captainquill** [comment]
Thanks for implementing opengl... I am having some rendering issues while upscaling. I checked with multiple games. Attaching the screenshot. 



![Screenshot_20240107_121201_melonDS Nightly](https://github.com/rafaelvcaetano/melonDS-android/assets/111143926/a0a289a0-0ce1-4b31-88f8-591c53e8ac92)

**Comment by ghost** [comment]
Just submitted this issue [here](https://github.com/rafaelvcaetano/melonDS-android/issues/1273)

it seems to continuously draw this mess overtop of models and you can sometimes see the model underneath.

![Screenshot_20240113-154242](https://github.com/rafaelvcaetano/melonDS-android/assets/14182708/0783d84b-f9b0-438d-a5a6-c63805e4a8e8)

**Comment by ghost** [comment]
> The OpenGL renderer has finally been introduced and, with it, comes the most requested feature since melonDS for Android began its development: **internal resolution scaling** or **high resolution rendering**. This allows the emulator to render the content using a resolution higher than the Nintendo DS's measly 256x192 resolution (per screen), which will give much more detail to 3D objects and scenes. To enable it, go to the Video settings and select the OpenGL renderer. If you can't see this option, it means that your device does not support OpenGL ES 3.2, which is required. You can scale the internal resolution up to 8 times, but I really doubt your device will handle it properly.
> 
> Here you can see a comparison between the software renderer and the OpenGL renderer running at 3x scaling:
> 
> Software	OpenGL (3x)
> ![image](https://private-user-images.githubusercontent.com/9086379/293907076-8bf78173-955e-4045-bb21-9d5a4fcd1ecb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUxNzU0ODksIm5iZiI6MTcwNTE3NTE4OSwicGF0aCI6Ii85MDg2Mzc5LzI5MzkwNzA3Ni04YmY3ODE3My05NTVlLTQwNDUtYmIyMS05ZDVhNGZjZDFlY2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTNUMTk0NjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTFjZjA5YWRlNDBhZWYwMWNlZWM3NTNmOTM0NjgxY2MyODhmNjVlM2U0MDZjOGJmYzk2NDk0ZTIwNWJhOTBjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.hl2CoN9UbCq95-OX5YguwtyI8hJIwUl7QJB1_pIjN78)	![image](https://private-user-images.githubusercontent.com/9086379/293907027-5f000600-59c3-49ab-b49e-e721cdc5a703.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUxNzU0ODksIm5iZiI6MTcwNTE3NTE4OSwicGF0aCI6Ii85MDg2Mzc5LzI5MzkwNzAyNy01ZjAwMDYwMC01OWMzLTQ5YWItYjQ5ZS1lNzIxY2RjNWE3MDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTNUMTk0NjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjdjMGExYjBjNjcwMTc5ZjQ5Y2IwZTJkMzI3NGE0MDQ3MjE0MjU4NTY1ZTdlMTFkZDM1MmI1MTA0ZjkwOTRlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.v1I-lqKHNjAG3C5ZIQEIRjM56zP69cuGXTYntL2Hfd8)
> ## Performance
> At 1x scaling, the performance is extremely similar to the software renderer. As you increase the resolution scaling, performance slightly goes down. To use use the highest scaling levels with acceptable performance, you will probably need a current gen flagship. But I don't have one, so I couldn't really test.
> 
> The current implementation is, above all, used as a proof that the renderer can run on Android. Some shortcuts were taken, which means that performance can still be slightly improved.
> 
> ## Rendering issues
> Just like in the PC version of melonDS, the OpenGL renderer is less accurate that the software one. As such, you are bound to find some visual artefacts or issues in some games. If you find such an issue, please take a look at [melonDS's OpenGL issues](https://github.com/melonDS-emu/melonDS/issues?q=is%3Aissue+is%3Aopen+label%3AOpenGL) to check if the PC version is affected by it as well. If it can't be reproduced on the PC version, feel free to open tickets so that I can take a look at them.
> 
> ## How to test
> Download the latest [nightly build](https://github.com/rafaelvcaetano/melonDS-android/releases/tag/nightly-release), enable the OpenGL render in the Video settings, and you are good to go! Don't forget to leave your feedback in this discussion!

@rafaelvcaetano if you have a discord server or such i would be glad to contact you and do thorough testing.

**Comment by HJMBgit84** [comment]
It is awesome... Keep going the great job that you are doing with this emulator, some games have graphical issues but I trust they will be fixed through the time, thanks for doing melonds on Android step up to the next level 🥳

**Comment by C-512L** [comment]
I was going to open an issue regarding OpenGL acceleration but it might better fit here. Maybe this project could benefit from using [libadrenotools](https://github.com/bylaws/libadrenotools) to allow loading external drivers like the open source mesa freedreno driver. I have seen a few emulators (like citra or yuzu) do it to fix broken driver bugs, which might be interesting considering how bad android GLES drivers can be.

**Comment by Unknowncat1991** [comment]
Opengl performance on my phone redmi note 9 is good with few minimal graphics issue when upscaling the resolution and a few frame drops. 

Does the opengl is multithreaded?
